[MRG + 1] Move n_iter and get_params invariance tests to common estimator_checks

JungeAlexander opened the pull request:
#### What does this implement/fix? Explain your changes.

Partially fixes issue #7533 by moving  tests relating to n_iter and invariance of get_params to common estimator_checks.
#### Any other comments?

Right now, test_transformer_n_iter() in utils/estimator_checks.py fails for some estimators that are not transformers. The reason is most likely that the test is now run for some estimators for which it is not applicable. This still needs to be fixed, see: https://github.com/scikit-learn/scikit-learn/commit/f534c3409b1a1e524e7c1e632c93b31d223d578e

Commit:
Test get_params invariance in common estimator tests

Remove test_get_params_invariance() from `test_common.py` and add
test call to _yield_all_tests() in `estimator_checks.py` to make
sure that get_params(deep=False) of a given Estimator returns a
subset of get_params(deep=True).

Compared to test_get_params_invariance(), it is NOT tested anymore
whether the given Estimator has an attribute get_params since
class BaseEstimator in `base.py` defines such an attribute
for each Estimator.

Partially addresses issue #7533
Also related to issue #4465

Commit:
Move test_transformer_n_iter() to estimator_checks.py

Remove the test test_transformer_n_iter() from tests/test_common.py
and perform the test logic in utils/estimator_checks.py instead.
Specifically, the method _yield_transformer_checks() now yields
check_transformer_n_iter() as part of the set of tests for
transformers.

test_transformer_n_iter() tests that that transformers with an
attribute max_iter, return the attribute of n_iter at least 1.

Partially addresses latter part of issue #7533

Commit:
Move test_non_transformer_estimators_n_iter() to estimator_checks.py

Remove the test_non_transformer_estimators_n_iter() from
tests/test_common.py; perform the test logic in
utils/estimator_checks.py instead.
Specifically, the method _yield_non_meta_checks() now yields
check_non_transformer_estimators_n_iter().

test_transformer_n_iter() tests that that estimators that are not
transformers with an attribute max_iter, return the attribute n_iter
of at least 1.

NOTE: The current implementation makes said test run for more
estimators than before this commit.
For some of these estimators, the test fails. This needs to be addressed
(see FIXME in line 111-115 of utils/estimator_checks.py for a potential
place to start).

Partially addresses latter part of issue #7533

Commit:
Fix check_non_transformer_estimators_n_iter calls

test_transformer_n_iter() test is now only run for
estimators where the test is applicable.

Partially addresses latter part of issue #7533

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1472,51 +1484,78 @@ def multioutput_estimator_convert_y_2d(name, y):
+                            'LinearSVC', 'LogisticRegression']
+
+    # Models made for multi-class scenarios
+    not_run_check_n_iter += ['MultiTaskElasticNet', 'MultiTaskElasticNetCV',
```
These should work. There's a ``if multi_output`` below just for these. There's also a ``make_multioutput_y`` helper.

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -186,6 +189,13 @@ def _yield_transformer_checks(name, Transformer):
         yield check_transformers_unfitted
+    # Dependent on external solvers and hence accessing the iter
+    # param is non-trivial.
+    external_solver = ['Isomap', 'KernelPCA', 'LocallyLinearEmbedding',
```
Hm I guess these were skipped manually before, so it's ok to skip them like this for this PR. It's really bad practice but I I think the PR is a step in the right direction.

JungeAlexander replied:
Yes, they were skipped before. I agree that finding another solution sounds like a great idea for the future.

amueller commented:
pep8 failure, otherwise this looks pretty ok.

Commit:
Run check_non_transformer_estimators_n_iter on multi-class estimators

To do this, use helper method multioutput_estimator_convert_y_2d.
Also remove multi_output parameter from
check_non_transformer_estimators_n_iter since this parameter is not
used anywhere and corresponding cases should be handled by said
helper method.

Also, some pep8 line length fixes.

jnothman commented:
Is this still a WIP or is it ready for full review?

JungeAlexander commented:
Should be ready for review now since all tests passed. Just renamed to [MRG]... to indicate this.

amueller commented:
LGTM apart from inline comment.

JungeAlexander commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1472,51 +1484,73 @@ def multioutput_estimator_convert_y_2d(name, y):
-    if name == 'AffinityPropagation':
-        estimator.fit(X)
+    # LassoLars stops early for the default alpha=1.0 the iris dataset.
+    if name == 'LassoLars':
```
Good point. This check was performed prior to the PR and I just moved it from `tests/test_common.py` to `utils/estimator_checks.py`: https://github.com/scikit-learn/scikit-learn/pull/7677/commits/f534c3409b1a1e524e7c1e632c93b31d223d578e#diff-b6e2c29089c86706dc868c7bfd58c1d6L174

If this is not done, the respective test fails:

```
======================================================================
FAIL: /home/ajunge/Code/scikit-learn/sklearn/tests/test_common.py.test_non_meta_estimators:check_non_transformer_estimators_n_iter(LassoLars)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ajunge/local/anaconda3/envs/sklearn/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/ajunge/Code/scikit-learn/sklearn/utils/testing.py", line 830, in __call__
    return self.check(*args, **kwargs)
  File "/home/ajunge/Code/scikit-learn/sklearn/utils/testing.py", line 355, in wrapper
    return fn(*args, **kwargs)
  File "/home/ajunge/Code/scikit-learn/sklearn/utils/estimator_checks.py", line 1522, in check_non_transformer_estimators_n_iter
    assert_greater_equal(estimator.n_iter_, 1)
  File "/home/ajunge/Code/scikit-learn/sklearn/utils/testing.py", line 156, in assert_greater_equal
    assert a >= b, message
AssertionError: 0 is not greater than or equal to 1
>>  assert 0 >= 1, '0 is not greater than or equal to 1'
```

Do you suggest this could be handled differently?

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1472,51 +1484,73 @@ def multioutput_estimator_convert_y_2d(name, y):
-        y_ = [[0.1, -0.2], [0.9, 1.1], [0.1, -0.5], [0.3, -0.2]]
+def check_transformer_n_iter(name, Estimator):
+    # Test that transformers with an attribute max_iter, return the attribute
+    # of n_iter at least 1.
```
*`n_iter_`

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1472,51 +1484,73 @@ def multioutput_estimator_convert_y_2d(name, y):
-        X = [[0., 0., 1.], [1., 0., 0.], [2., 2., 2.], [2., 5., 4.]]
-        y_ = [[0.1, -0.2], [0.9, 1.1], [0.1, -0.5], [0.3, -0.2]]
+def check_transformer_n_iter(name, Estimator):
+    # Test that transformers with an attribute max_iter, return the attribute
```
"attribute max_iter" -> "parameter max_iter"

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1472,51 +1484,73 @@ def multioutput_estimator_convert_y_2d(name, y):
-        y_ = np.reshape(y_, (-1, 1))
+def check_non_transformer_estimators_n_iter(name, Estimator):
+    # Test that estimators that are not transformers with an attribute
+    # max_iter, return the attribute of n_iter at least 1.
```
*`n_iter_`

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1472,51 +1484,73 @@ def multioutput_estimator_convert_y_2d(name, y):
-    if multi_output:
-        y_ = np.reshape(y_, (-1, 1))
+def check_non_transformer_estimators_n_iter(name, Estimator):
+    # Test that estimators that are not transformers with an attribute
```
"attribute" -> "parameter"

jnothman reviewed (changes requested):
Otherwise this is looking good.

Commit:
Fix documentation for n_iter tests

There was some confusion between attributes and parameters.
Also rename n_iter to n_iter_

JungeAlexander commented:
Thanks for the feedback. Issues should have been fixed: https://github.com/scikit-learn/scikit-learn/pull/7677/commits/d93a10d093cfc9afb76acad026ffb9d44ff911ac

amueller commented:
thanks :)

amueller commented:
I wonder if this needs a whatsnew, because it added another check to ``check_estimator`` - and one that currently makes use of a lot of hard-coded classes...
